### PR TITLE
#149. Add http scheduler tests

### PR DESCRIPTION
**Pull request by avbelov23**

### CODE REVIEW

File: tempesta_fw/cfg.c [hunk 1]
```diff
@@ -848,6 +848,10 @@ spec_cleanup(TfwCfgSpec specs[])
 			spec->cleanup(spec);
 		}
 		spec->call_counter = 0;
+
+		if (spec->handler == &tfw_cfg_handle_children) {
```
Comment (i-rinat) [inline tempesta_fw/cfg.c]: Looks like solving some issue. It would be nice to have that in a separate commit.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 2]
```diff
@@ -189,6 +192,8 @@ TEST(tfw_sched_hash, max_srv_in_sg_and_max_conn)
 
 TEST_SUITE(sched_hash)
 {
+	tfw_http_parse_req_ptr = get_sym_ptr("tfw_http_parse_req");
+
```
Comment (i-rinat) [inline tempesta_fw/t/unit/test_sched_hash.c]: A `BUG_ON(tfw_http_parse_req_ptr == NULL);` here will help to catch `NULL` as early as possible.

File: tempesta_fw/t/unit/kallsyms_helper.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+#include <linux/string.h>
+#include <linux/kallsyms.h>
+#include <linux/module.h>
+
+#include "kallsyms_helper.h"
+
+typedef struct {
+        unsigned long addr;
```
Comment (i-rinat) [inline tempesta_fw/t/unit/kallsyms_helper.c]: spaces vs. tabs.

File: tempesta_fw/t/unit/kallsyms_helper.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        const char    *name;
+} Symdata;
+
+static int
+get_sym(void *data, const char *namebuf, struct module *owner, unsigned long addr)
+{
+        Symdata *symdata = (Symdata *)data;
```
Comment (i-rinat) [inline tempesta_fw/t/unit/kallsyms_helper.c]: no explicit casting is required here, `(void *)` is implicitly converted to any other pointer.

File: tempesta_fw/t/unit/test_sched_http.c [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+cleanup_cfg(void)
+{
+	TfwCfgMod cfg_mod;
+
+	cfg_mod = *tfw_cfg_mod_lookup("tfw_sched_http");
+	spec_cleanup_ptr(cfg_mod.specs);
+}
+
+static void
+test_req(char *req_str, TfwSrvConnection *expect_conn)
```
Comment (i-rinat) [inline tempesta_fw/t/unit/test_sched_http.c]: `req_str` is of type `char *` here, but some calls of `test_req` later use string literals, which are in fact, `const`.
Since `tfw_http_parse_req` can write to `req_str`, that can fail.

File: tempesta_fw/t/unit/test_sched_http.c [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	                                   match sg9 hdr_raw eq User-Agent:Bot;\n\
+	                                   match sg10 hdr_raw prefix X-Forwarded-For;\n}\n")) {
+		TEST_FAIL("can't parse rules\n");
+	}
+
+	test_req("GET http://natsys-lab.com/foo HTTP/1.1\r\n\r\n", expect_conn1);
+	test_req("GET http://natsys-lab.com/foo/bar/ HTTP/1.1\r\n\r\n", expect_conn2);
+	test_req("GET http://natsys-lab.com/foo/baz/ HTTP/1.1\r\n\r\n", expect_conn3);
+	test_req("GET http://natsys-lab2.com/foo/baz/ HTTP/1.1\r\n\r\n", expect_conn4);
+	test_req("GET http://google.com/foo/baz/ HTTP/1.1\r\nHost: google.com\r\n", expect_conn5);
```
Comment (i-rinat) [inline tempesta_fw/t/unit/test_sched_http.c]: Some requests are not complete (only one pair of `\r\n` at the end).

Comment (i-rinat) [inline tempesta_fw/t/unit/test_sched_http.c]: Some lines are longer than 80 character positions.

File: tempesta_fw/t/unit/test_sched_http.c [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+		.bad_req_str = "GET http://natsys-lab.com/foo HTTP/1.1\r\nConnection: close\r\n",
+	},
+	{
+		.rule_str = "sched_http_rules {\nmatch default hdr_raw eq User-Agent:Bot;\n}\n",
+		.good_req_str = "GET http://natsys-lab.com/foo HTTP/1.1\r\nUser-Agent:Bot\r\n",
+		.bad_req_str = "GET http://natsys-lab.com/foo HTTP/1.1\r\nUser-Agent:Tot\r\n",
+	},
+	{
+		.rule_str = "sched_http_rules {\nmatch default hdr_raw prefix User-Agent;\n}\n",
+		.good_req_str = "GET http://natsys-lab.com/foo HTTP/1.1\r\nUser-Agent:Bot\r\n",
```
Comment (i-rinat) [inline tempesta_fw/t/unit/test_sched_http.c]: Field names are case-insensitive. Maybe it's worth to test mismatching character cases in rule and requests too?

Comment (avbelov23) [inline tempesta_fw/t/unit/test_sched_http.c]: For this there are some more in-depth tests (test_http_match.c)

File: tempesta_fw/t/unit/test_sched_http.c (hunk 6 above)
Comment (avbelov23) [inline tempesta_fw/t/unit/test_sched_http.c]: > Some lines are longer than 80 character positions.
This is does not increase readability.
(https://raw.githubusercontent.com/krizhanovsky/NatSys-Lab/master/CodingStyle)

File: tempesta_fw/t/unit/test_sched_http.c (hunk 5 above)
Comment (avbelov23) [inline tempesta_fw/t/unit/test_sched_http.c]: Okey.

File: tempesta_fw/t/unit/kallsyms_helper.c (hunk 4 above)
Comment (avbelov23) [inline tempesta_fw/t/unit/kallsyms_helper.c]: Okey. Legacy of C ++.

File: tempesta_fw/t/unit/kallsyms_helper.c (hunk 3 above)
Comment (avbelov23) [inline tempesta_fw/t/unit/kallsyms_helper.c]: Really! Accidentally Indent Using Spaces flag for the file.

File: tempesta_fw/t/unit/test_sched_hash.c (hunk 2 above)
Comment (avbelov23) [inline tempesta_fw/t/unit/test_sched_hash.c]: Okey.

File: tempesta_fw/cfg.c (hunk 1 above)
Comment (avbelov23) [inline tempesta_fw/cfg.c]: Perhaps enough to comment?
